File: 0083.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        cur = ListNode(0)
        res = cur
        cur.next = head
        cur = cur.next
        while head:
            if cur.val == head.val:
                if head.next == None:
                    cur.next = None
                head = head.next
            else:
                cur.next = head
                cur = cur.next
                head = head.next
        return res.next


class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        cur = head
        while cur and cur.next:
            if cur.val == cur.next.val:
                cur.next = cur.next.next
            else:
                cur = cur.next
        return head

File: 0083.cpp
// First time
/*
    T: O(n)/S: O(1)
*/
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        ListNode *cur = head;
        while (cur && cur->next) {
            if (cur->val == cur->next->val) {
                cur->next = cur->next->next;
            } else {
                cur = cur->next;
            }
        }
        return head;
    }
};

class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (!head)
            return head;
        ListNode *ret = new ListNode(head->val), *cur = ret;
        while (head) {
            if (head->val != cur->val) {
                cur->next = new ListNode(head->val);
                cur = cur->next;
            }
            head = head->next;
        }
        return ret;
    }
};